File: templates/graph/graph.go
package graph

func BFSLevels(graph map[int][]int, start int) []int {
	queue := []int{start}
	dist := map[int]int{start: 0}
	order := []int{}

	for len(queue) > 0 {
		node := queue[0]
		queue = queue[1:]
		order = append(order, node)

		for _, nei := range graph[node] {
			if _, seen := dist[nei]; seen {
				continue
			}
			dist[nei] = dist[node] + 1
			queue = append(queue, nei)
		}
	}
	return order
}

func DFSRecursive(graph map[int][]int, start int) []int {
	seen := map[int]bool{}
	order := []int{}

	var dfs func(int)
	dfs = func(node int) {
		seen[node] = true
		order = append(order, node)
		for _, nei := range graph[node] {
			if !seen[nei] {
				dfs(nei)
			}
		}
	}

	dfs(start)
	return order
}

func NumIslands(grid [][]byte) int {
	rows, cols := len(grid), len(grid[0])
	count := 0

	var dfs func(int, int)
	dfs = func(r, c int) {
		if r < 0 || r >= rows || c < 0 || c >= cols || grid[r][c] != '1' {
			return
		}
		grid[r][c] = '0'
		dfs(r+1, c)
		dfs(r-1, c)
		dfs(r, c+1)
		dfs(r, c-1)
	}

	for r := 0; r < rows; r++ {
		for c := 0; c < cols; c++ {
			if grid[r][c] == '1' {
				count++
				dfs(r, c)
			}
		}
	}

	return count
}

File: templates/graph/graph.py
from collections import deque


def bfs_levels(graph: dict[int, list[int]], start: int) -> list[int]:
    queue = deque([start])
    dist = {start: 0}
    order: list[int] = []

    while queue:
        node = queue.popleft()
        order.append(node)
        for nei in graph.get(node, []):
            if nei in dist:
                continue
            dist[nei] = dist[node] + 1
            queue.append(nei)

    return order


def dfs_recursive(graph: dict[int, list[int]], start: int) -> list[int]:
    seen: set[int] = set()
    order: list[int] = []

    def dfs(node: int) -> None:
        seen.add(node)
        order.append(node)
        for nei in graph.get(node, []):
            if nei not in seen:
                dfs(nei)

    dfs(start)
    return order


def num_islands(grid: list[list[str]]) -> int:
    rows, cols = len(grid), len(grid[0])
    count = 0

    def dfs(row: int, col: int) -> None:
        if row < 0 \
            or row >= rows \
            or col < 0 \
            or col >= cols \
            or grid[row][col] != "1":
            return
        grid[row][col] = "0"
        dfs(row + 1, col)
        dfs(row - 1, col)
        dfs(row, col + 1)
        dfs(row, col - 1)

    for row in range(rows):
        for col in range(cols):
            if grid[row][col] == "1":
                count += 1
                dfs(row, col)

    return count
